core/src/main/scala/org/apache/spark/SparkException.scala

```diff
@@ -30,3 +30,13 @@ class SparkException(message: String, cause: Throwable)
  */
 private[spark] class SparkDriverExecutionException(cause: Throwable)
   extends SparkException("Execution error", cause)
+
+/**
+ * Exception indicating an error internal to Spark -- it is in an inconsistent state, not due
+ * to any error by the user
+ */
+class SparkIllegalStateException(message: String, cause: Throwable)
```

Comment:
I'd prefer to stick with the existing strategy, which is to use IllegalStateExceptions and introduce this new exception in a separate PR (my understanding is that we do currently use IllegalStateExceptions only for things that aren't the user's fault; if not, agree with your sentiment that it would be nice to clean that up, but not as part of this PR).